### ISSUE TITLE
🎛️ Migrate button longrunning a Stimulus Controller

### ISSUE THREAD

**Issue by lb-** [wagtail/wagtail#9910]
> ℹ️ **Part of the [Stimulus  🎛️ RFC 78](https://github.com/wagtail/rfcs/pull/78)**

### Is your proposal related to a problem?

We have a core.js implementations of JavaScript code that, when a button is clicked, makes the button disabled and then shows a spinner icon on the button for a fixed duration.

### Describe the solution you'd like

* Create a stimulus controller `w-loader` that will replace the existing ad-hoc JS implementation
* The behaviour should be exactly the same as current state but using Stimulus data attributes for the behaviour & classes declaration
* Controller should be written in TypeScript
* Ensure that the existing unit tests are created to reflect this new behaviour
* Ensure that there is a Storybook story for this component
* We will need to document an upgrade consideration that the previous classes only approach will no longer work
* We will need to ensure that existing `button-longrunning` only still works as expected (this will use https://stimulus.hotwired.dev/reference/controllers#trigger-behaviour-when-a-controller-is-registered )
   * Important: this will need further discussion.

### Additional context

* Implementation https://github.com/wagtail/wagtail/blob/main/client/src/entrypoints/admin/core.js#L310-L372

#### Current usage

```
<button
  type="submit"
  class="button button-longrunning"
  data-clicked-text="{% trans 'Signing in…' %}"
>
  {% icon name="spinner" %}
  <em>{% trans 'Sign in' %}</em>
</button>
// adds button-longrunning-active when active
// checks form validation if inside a form
```

#### Potential usage

```
<button
  type="submit"
  class="button button-longrunning"
  data-controller="w-loader"
  data-w-loader-active-class="button-longrunning-active"
  data-w-loader-active-value="{% trans 'Signing in…' %}" // could be nothing, so need to think of defaults here
  data-w-loader-duration-seconds-value="40" // defaults to 30 (seconds) - may want to use milliseconds instead
  data-action="w-loader#activate" // or start or trigger?
>
  {% icon name="spinner" %}
  <em data-w-loader-target="label">{% trans 'Sign in' %}</em> // want to be able to fallback on `em` if target is not present
</button>
```

#### See also

* https://github.com/wagtail/wagtail/issues/8232

**Comment by lb-** [comment]
As per our call @Lovelyfin00 

```ts
 static afterLoad(
    identifier: string /* w-load-button */,
    application: Application,
  ) {
    const { controllerAttribute } = application.schema;
    // https://css-tricks.com/almanac/selectors/a/attribute/ (read about data attribute selectors)
    document
      .querySelectorAll(
        `.button-longrunning:not([data-controller~='${identifier}'])`,
      )
      .forEach((button) => {
        console.log(
          "YO! this button needs updatin' - this way wont work after Wagtail 5.0",
          button,
        );
        button.setAttribute('data-controller', identifier);
        // adds data-controller="w-loader"
        // we will need to add other attributes like data-action
      });

    // this is NOT the instance (a specific button usage), but the controller itself
    console.log('afterLoad - run ONCE - when the controller is registered', {
      identifier,
      application,
      controllerAttribute,
      schema: application.schema, // tells stimulus what attributes are used for action, controller etc
    });
  }
```

**Comment by lb-** [comment]
@Lovelyfin00  - once #9956 is merged, please can you action the following next items in a new PR.

* [ ] Change the `duration-seconds` to simply `duration` and use milliseconds - I think that there would be reasonable use cases for this to be under one second, we have the ability to set the default so this 'long' value (`30000`) would only be shown if someone intentionally wants something different.
* [ ] Update all existing usage of `.button-longrunning` to the new data attributes across the entire admin (remembering we have default attributes and only add the data attributes that are required to match existing behaviour)
* [ ] We may want to allow the `em` to be used without the data target attribute, you would need to handle this behaviour in the `connect` method, in fact you can just pull out the code you wrote in the `afterLoad` that does this - find the inside `em` and add the data target attribute IF there is no target.
* [ ] Add an upgrade consideration note to the 5.0 release notes (advising of the old data attributes and the new ones)
* [ ] Update the styleguide examples to have a custom duration (5 seconds) & change that button's text to something like (click me 5s)

Amazing stuff here - after a bit of polish, we can close this and also close  https://github.com/wagtail/wagtail/issues/8232

**Comment by Lovelyfin00** [comment]
@lb- 
So how long do you want me to set the duration 10 seconds? I feel 1 second would be too short

**Comment by lb-** [comment]
@Lovelyfin00  yep that's it.

Imagine there is some html as follows
```
<button class="button-longrunning" data-controller="w-action" data-action="w-action#post" type="button"><em>Some label</em></button>
```

The current `afterLoad` will 'wipe out' the existing data-controller/action attributes, so the existing behaviour will no longer work.

```
<button class="button-longrunning" data-controller="w-progress" data-action="w-progress#activate" type="button"><em data-w-progress-target="label">Some label</em></button>
```

Note that the simplest way to do this is something like the following (double check this syntax).

```js
someElement.setAttribute('data-action', (someElement.getAttribute('data-action') || '') + 'new-thing');
```

This is quite verbose I know, likely if we have to do this kind of thing a lot we will pull this out to a util, this is where using vanilla JS is sometimes a bit more annoying than using jQuery.

You can use the `dataset` syntax to make it a bit nicer but that requires using a camelCase version of the data attributes.

https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset#name_conversion

**Comment by Lovelyfin00** [comment]
> [ ] We may want to allow the `em` to be used without the data target attribute, you would need to handle this behaviour in the `connect` method, in fact you can just pull out the code you wrote in the `afterLoad` that does this - find the inside `em` and add the data target attribute IF there is no target.

I don't understand this part. Do you mean that I should remove the `data-w-progress-target="label"` and check if the element has an `em` as it's child and if it does, I should attach the activaValue in the `connect()` method?

**Comment by lb-** [comment]
@Lovelyfin00 you're on the right track. 

Remove this code from `afterLoad`

```
const labelElement = button.querySelector('em');
            if (labelElement) {
              labelElement.setAttribute(`data-${identifier}-target`, 'label');
            }
```

And add a `connect` method with this code (roughly)

```
connect() {
  if (this.hasLabelTarget) return;
const labelElement = this.element.querySelector('em');
  if (!labelElement) return;  
   labelElement.setAttribute(`data-${identifier}-target`, 'label');
}
```

Essentially the same behaviour, but in the connect method instead. This way developers don't have to add the target attributes if they have an `em`. 

Note. Our core code should still add the data attributes but this just makes it a little more convenient to adopt.

**Comment by Lovelyfin00** [comment]
Ok great
I understand now

**Comment by lb-** [comment]
Turns out this Stimulus adoption fixed an unknown bug - 0187a1494bc56bab4592dc0d3d314121edf590d0 (release notes added).

The progress button is used in the dynamically injected image/document update form (multiple uploads), utill know the JS  never ran for this as it was added async.

Now that Stimulus is watching the DOM - the behaviour works as expected for these buttons.

<img width="1570" alt="Screenshot 2023-02-16 at 8 21 53 am" src="https://user-images.githubusercontent.com/1396140/219196114-326264fb-dea3-4a28-8bc5-6647aac53801.png">

Thanks Loveth!

**Comment by Lovelyfin00** [comment]
Oh wow
So the button-long running never worked here before?
I tested this page when I added the data attributes to the buttons
This is great🎉🎉

**Comment by lb-** [comment]
Yeah. I've never seen it go disabled on update until now - so the html would have been there all along but because the init function only ran once on DOM load it never picked up html added dynamically.
